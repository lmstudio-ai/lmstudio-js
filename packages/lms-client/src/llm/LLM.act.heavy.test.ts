import { z } from "zod";
import {
  type ChatMessage,
  type LLM,
  LMStudioClient,
  type PredictionResult,
  tool,
} from "../index.js";
import { ensureHeavyTestsEnvironment, llmTestingQwen05B } from "../shared.heavy.test.js";

describe("LLM.act", () => {
  let client: LMStudioClient;
  let model: LLM;
  const addImplementation = jest.fn(({ a, b }) => a + b);
  const additionTool = tool({
    name: "add",
    description: "Add two numbers",
    parameters: { a: z.number(), b: z.number() },
    implementation: addImplementation,
  });
  beforeAll(async () => {
    client = new LMStudioClient();
    await ensureHeavyTestsEnvironment(client);
  });
  beforeEach(async () => {
    model = await client.llm.model(llmTestingQwen05B, {
      verbose: false,
      config: {
        llamaKCacheQuantizationType: "f32",
        llamaVCacheQuantizationType: "f32",
      },
    });
  }, 60_000);
  it("should call the tool with the correct parameters", async () => {
    const onMessage = jest.fn();
    const onFirstToken = jest.fn();
    const onPredictionCompleted = jest.fn();
    const onPredictionFragment = jest.fn();
    const onPromptProcessingProgress = jest.fn();
    const onRoundStart = jest.fn();
    const onRoundEnd = jest.fn();

    await model.act('First say "Hi". Then calculate 1 + 3 with the tool.', [additionTool], {
      temperature: 0,
      onMessage,
      onFirstToken,
      onPredictionCompleted,
      onPredictionFragment,
      onPromptProcessingProgress,
      onRoundStart,
      onRoundEnd,
    });
    expect(addImplementation).toHaveBeenCalledTimes(1);
    expect(addImplementation.mock.calls[0]).toEqual([{ a: 1, b: 3 }]);
    expect(onMessage).toHaveBeenCalledTimes(3);

    const message0 = onMessage.mock.calls[0][0] as ChatMessage;
    expect(message0.getRole()).toBe("assistant");
    expect(message0.getText()).toContain("Hi");
    const message1 = onMessage.mock.calls[1][0] as ChatMessage;
    expect(message1.getRole()).toBe("tool");
    expect(message1.getText()).toBe("");
    const message2 = onMessage.mock.calls[2][0] as ChatMessage;
    expect(message2.getRole()).toBe("assistant");
    expect(message2.getText()).toContain("4");

    expect(onFirstToken.mock.calls).toEqual([[0], [1]]);

    expect(onPredictionCompleted).toHaveBeenCalledTimes(2);
    const prediction0 = onPredictionCompleted.mock.calls[0][0] as PredictionResult;
    expect(prediction0.content).toContain("Hi");
    expect(prediction0.roundIndex).toEqual(0);
    expect(prediction0.modelInfo).toMatchSnapshot({
      identifier: expect.any(String),
      instanceReference: expect.any(String),
      modelKey: expect.any(String),
    });
    expect(prediction0.stats).toMatchSnapshot({
      numGpuLayers: expect.any(Number),
      timeToFirstTokenSec: expect.any(Number),
      tokensPerSecond: expect.any(Number),
    });
    const prediction1 = onPredictionCompleted.mock.calls[1][0] as PredictionResult;
    expect(prediction1.content).toContain("4");
    expect(prediction1.roundIndex).toEqual(1);
    expect(prediction1.modelInfo).toMatchSnapshot({
      identifier: expect.any(String),
      instanceReference: expect.any(String),
      modelKey: expect.any(String),
    });
    expect(prediction1.stats).toMatchSnapshot({
      numGpuLayers: expect.any(Number),
      timeToFirstTokenSec: expect.any(Number),
      tokensPerSecond: expect.any(Number),
    });

    // Cannot assert on content due to non-determinism
    expect(onPredictionFragment).toHaveBeenCalled();

    expect(onPromptProcessingProgress).toHaveBeenCalledWith(0, 0);
    expect(onPromptProcessingProgress).toHaveBeenCalledWith(0, 1);
    expect(onPromptProcessingProgress).toHaveBeenCalledWith(1, 0);
    expect(onPromptProcessingProgress).toHaveBeenCalledWith(1, 1);

    expect(onRoundStart).toHaveBeenCalledTimes(2);
    expect(onRoundStart.mock.calls).toEqual([[0], [1]]);

    expect(onRoundEnd).toHaveBeenCalledTimes(2);
    expect(onRoundEnd.mock.calls).toEqual([[0], [1]]);
  });
});
