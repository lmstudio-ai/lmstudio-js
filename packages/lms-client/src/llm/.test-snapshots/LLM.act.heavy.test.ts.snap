// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LLM.act should call the tool with the correct parameters 1`] = `
{
  "architecture": "qwen2",
  "contextLength": 4096,
  "displayName": "Qwen2.5 0.5B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "0.5B",
  "path": "lmstudio-community/Qwen2.5-0.5B-Instruct-GGUF/Qwen2.5-0.5B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 397807936,
  "trainedForToolUse": true,
  "type": "llm",
  "vision": false,
}
`;

exports[`LLM.act should call the tool with the correct parameters 2`] = `
{
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 38,
  "promptTokensCount": 192,
  "stopReason": "eosFound",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalTokensCount": 230,
}
`;

exports[`LLM.act should call the tool with the correct parameters 3`] = `
{
  "architecture": "qwen2",
  "contextLength": 4096,
  "displayName": "Qwen2.5 0.5B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "0.5B",
  "path": "lmstudio-community/Qwen2.5-0.5B-Instruct-GGUF/Qwen2.5-0.5B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 397807936,
  "trainedForToolUse": true,
  "type": "llm",
  "vision": false,
}
`;

exports[`LLM.act should call the tool with the correct parameters 4`] = `
{
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 13,
  "promptTokensCount": 249,
  "stopReason": "eosFound",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalTokensCount": 262,
}
`;
